workloads/wmt/wmt_jax/train.py

```diff
@@ -0,0 +1,381 @@
+"""Machine Translation example.
+
+This script trains a Transformer on a WMT dataset.
+"""
+
+# pytype: disable=wrong-arg-count
+# pytype: disable=attribute-error
+
+import collections
+
+from . import bleu
+from . import decode
+from . import models
+from absl import logging
+from flax import linen as nn
+from flax.training import common_utils
+import jax
+import jax.numpy as jnp
+import numpy as np
+import tensorflow as tf
+
+
+def create_learning_rate_scheduler(
+    factors="constant * linear_warmup * rsqrt_decay",
+    base_learning_rate=0.5,
+    warmup_steps=1000,
+    decay_factor=0.5,
+    steps_per_decay=20000,
+    steps_per_cycle=100000):
+  """Creates learning rate schedule.
+
+  Interprets factors in the factors string which can consist of:
+  * constant: interpreted as the constant value,
+  * linear_warmup: interpreted as linear warmup until warmup_steps,
+  * rsqrt_decay: divide by square root of max(step, warmup_steps)
+  * rsqrt_normalized_decay: divide by square root of max(step/warmup_steps, 1)
+  * decay_every: Every k steps decay the learning rate by decay_factor.
+  * cosine_decay: Cyclic cosine decay, uses steps_per_cycle parameter.
+
+  Args:
+    factors: string, factors separated by "*" that defines the schedule.
+    base_learning_rate: float, the starting constant for the lr schedule.
+    warmup_steps: int, how many steps to warm up for in the warmup schedule.
+    decay_factor: float, the amount to decay the learning rate by.
+    steps_per_decay: int, how often to decay the learning rate.
+    steps_per_cycle: int, steps per cycle when using cosine decay.
+
+  Returns:
+    a function learning_rate(step): float -> {"learning_rate": float}, the
+    step-dependent lr.
+  """
+  factors = [n.strip() for n in factors.split("*")]
+
+  def step_fn(step):
+    """Step to learning rate function."""
+    ret = 1.0
+    for name in factors:
+      if name == "constant":
+        ret *= base_learning_rate
+      elif name == "linear_warmup":
+        ret *= jnp.minimum(1.0, step / warmup_steps)
+      elif name == "rsqrt_decay":
+        ret /= jnp.sqrt(jnp.maximum(step, warmup_steps))
+      elif name == "rsqrt_normalized_decay":
+        ret *= jnp.sqrt(warmup_steps)
+        ret /= jnp.sqrt(jnp.maximum(step, warmup_steps))
+      elif name == "decay_every":
+        ret *= (decay_factor**(step // steps_per_decay))
+      elif name == "cosine_decay":
+        progress = jnp.maximum(0.0,
+                               (step - warmup_steps) / float(steps_per_cycle))
+        ret *= jnp.maximum(0.0,
+                           0.5 * (1.0 + jnp.cos(jnp.pi * (progress % 1.0))))
+      else:
+        raise ValueError("Unknown factor %s." % name)
+    return jnp.asarray(ret, dtype=jnp.float32)
+
+  return step_fn
+
+
+def compute_weighted_cross_entropy(logits,
```

Comment:
we should remove this function, as it seems like a duplicate of `loss_fn` in `workload.py`.

Reply:
I reused it in loss_fn